lib/hallux/ordered_map.ex

```diff
@@ -0,0 +1,242 @@
+  * All keys to an item can be retrieved in O(m*log(n)),
+    where m is the number of occurrences of that value.
+    (Currently, hash collisions are not checked.)
+  """
+
+  alias Hallux.Internal.FingerTree
+  alias Hallux.Internal.FingerTree.Empty
+  alias Hallux.Internal.OrderedHashBounds
+  alias Hallux.Internal.OrderedHashBounds.Order
+
+  defstruct [:finger_tree]
+
+  @doc """
+  `(O(1))`. Returns an empty `OrderedMap`.
+
+  ## Examples
+
+      iex> new()
+      #HalluxOrderedMap<[]>
+  """
+  def new() do
+    %__MODULE__{finger_tree: empty()}
+  end
+
+  @doc """
+  `(O(n log(n)))`. Returns an `OrderedMap` filled with the list of key-value-pairs.
+
+  ## Examples
+
+      iex> new([{1, "hai"}]) |> Enum.to_list()
+      [{1, "hai"}]
+      iex> new([{23, 5}, {1, "hai"}, {"something", [:some, :thing]}, {:foo, "foo"}])
+      #HalluxOrderedMap<[{1, "hai"}, {23, 5}, {:foo, "foo"}, {"something", [:some, :thing]}]>
+  """
+  def new(key_value_list) do
+    Enum.into(key_value_list, new())
+  end
+
+  #      {#HalluxOrderedMap<[a: "a"]>, #HalluxOrderedMap<[b: "b1", b: "b2"]>, #HalluxOrderedMap<[c: "c"]>}
+
+  @doc """
+  Split according to the order key.
+
+  ## Examples
+
+      iex> my_map = new([b: "b1", a: "a", c: "c", b: "b2"])
+      #HalluxOrderedMap<[a: "a", b: "b1", b: "b2", c: "c"]>
+      iex> {a, b, c} = split(my_map, :b)
+      iex> a
+      #HalluxOrderedMap<[a: "a"]>
+      iex> b
+      #HalluxOrderedMap<[b: "b1", b: "b2"]>
+      iex> c
+      #HalluxOrderedMap<[c: "c"]>
+  """
+  def split(%__MODULE__{finger_tree: finger_tree}, middle) do
+    {smaller, greater_or_equal} =
+      FingerTree.split(
+        finger_tree,
+        fn %OrderedHashBounds{order: order} ->
+          OrderedHashBounds.compare_order(order, %Order{order: middle}) in [:eq, :gt]
+        end
+      )
+
+    {equal, greater} =
+      FingerTree.split(
+        greater_or_equal,
+        fn %OrderedHashBounds{order: order} ->
+          OrderedHashBounds.compare_order(order, %Order{order: middle}) == :gt
+        end
+      )
+
+    {%__MODULE__{finger_tree: smaller}, %__MODULE__{finger_tree: equal},
+     %__MODULE__{finger_tree: greater}}
+  end
+
+  @doc """
+  Get an ordered map of all values with the given order key.
+
+    ## Examples
+
+      iex> my_map = new([b: "b1", a: "a", c: "c", b: "b2"])
+      #HalluxOrderedMap<[a: "a", b: "b1", b: "b2", c: "c"]>
+      iex> get(my_map, :b)
+      #HalluxOrderedMap<[b: "b1", b: "b2"]>
+  """
+  def get(ordered_hash_bounds_queue, middle) do
+    {_, equal, _} = split(ordered_hash_bounds_queue, middle)
+    equal
+  end
+
+  @doc """
+  Insert a value and key.
+
+  ## Examples
+
+      iex> my_map = new([b: "b1", a: "a", c: "c", b: "b2"])
+      #HalluxOrderedMap<[a: "a", b: "b1", b: "b2", c: "c"]>
+      iex> insert(my_map, :a, "A")
+      #HalluxOrderedMap<[a: "a", a: "A", b: "b1", b: "b2", c: "c"]>
+  """
+  def insert(ordered_hash_bounds_queue, middle, value) do
```

Comment:
suggestion: replace `middle` with `key`, to keep consistency with the doc.

Reply:
:+1: